### PR TITLE
RW 601: Widget with description

### PR DESCRIPTION
**Pull request by attiks**

### CODE REVIEW

File: composer.json
Comment (orakili) [inline composer.json:35]: RW has already an implementation to handle adding option attributes: https://github.com/UN-OCHA/rwint9-site/blob/f563644a643e80963d517beda2094fd58743a57d/html/modules/custom/reliefweb_form/reliefweb_form.module#L29-L30 using the `#option_attributes` property, that do something similar.

File: config/core.entity_form_display.node.training.default.yml
Comment (orakili) [inline config/core.entity_form_display.node.training.default.yml:188]: I don't see the code for the widget in the PR but if that's not already the case, please have a look at the https://github.com/UN-OCHA/rwint9-site/blob/develop/html/modules/custom/reliefweb_fields/src/Plugin/Field/FieldWidget/ReliefWebEntityReferenceSelect.php widget which adds widget options to add fields as data attributes to select option elements.

This widget is for `select` but could be adapted easily enough for `checkboxes/radios`.

It's used for example for the disaster field on the report form to add the disaster status as a data attribute which is then used to add a colored border to the suggestions.

File: composer.json
Comment (attiks) [reply composer.json]: Module removed

File: config/core.entity_form_display.node.training.default.yml
Comment (attiks) [reply config/core.entity_form_display.node.training.default.yml:188]: Missing file added

**Comment by attiks** [comment]
PS: Labels in popup can be clicked to control radfio/checkbox

File: config/core.extension.yml
Comment (orakili) [inline config/core.extension.yml:18]: Devel module is enabled via config_split: https://github.com/UN-OCHA/rwint9-site/blob/c41fd92c2e6ecb511bbb0001b616099d63a232ac/config/config_split.config_split.config_dev.yml#L14

File: composer.lock
Comment (orakili) [inline composer.lock:15301]: Doesn't matter much but the version of composer used for the current site images in 2.3.x.

File: html/modules/custom/reliefweb_fields/src/Plugin/Field/FieldWidget/ReliefWebOptions.php
Comment (orakili) [inline html/modules/custom/reliefweb_fields/src/Plugin/Field/FieldWidget/ReliefWebOptions.php:23]: The code seems to only work for entity reference field types so I think that should be removed.

File: html/modules/custom/reliefweb_fields/src/Plugin/Field/FieldWidget/ReliefWebOptions.php
Comment (orakili) [inline html/modules/custom/reliefweb_fields/src/Plugin/Field/FieldWidget/ReliefWebOptions.php:42]: The code below seems to only work for the taxonomy terms so maybe it can bail out earlier with something like:

```php
if ($entity_type_id !== 'taxonomy_term') {
   return parent::formElement($items, $delta, $element, $form, $form_state);
}
```

File: html/modules/custom/reliefweb_fields/src/Plugin/Field/FieldWidget/ReliefWebOptions.php
Comment (orakili) [inline html/modules/custom/reliefweb_fields/src/Plugin/Field/FieldWidget/ReliefWebOptions.php:55]: Media or nodes uses `type` not `bundle` for example. Also, as there is only one `entity_type_id`, maybe the whole foreach block can be simplified:


```php
$entities = $entity_type_manager->getStorage($entity_type_id)->loadByProperties([
  $this->getEntityTypeBundleField($entity_type_id) => $bundles,
]);
```

**Review by orakili — COMMENTED** [review]
Thanks, that seems to be working well.

The `css` and possibly the `js` file will probably need to be overridden in the subtheme to add the necessary classes so that the styling from the `rw-guidelines` component can be re-used.

Also the fact that clicking on the terms in the popup select the guidelines is quite confusing in terms of UX and accessibility as there is no feedback when doing so and the selection only appears after closing the popup. I think we should avoid that and just let the popup be help information without interactivity.

File: html/modules/custom/reliefweb_fields/src/Plugin/Field/FieldWidget/ReliefWebOptions.php
Comment (attiks) [reply html/modules/custom/reliefweb_fields/src/Plugin/Field/FieldWidget/ReliefWebOptions.php:23]: True

**Comment by attiks** [comment]
> Also the fact that clicking on the terms in the popup select the terms is quite confusing in terms of UX and accessibility as there is no feedback when doing so and the selection only appears after closing the popup. I think we should avoid that and just let the popup be help information without interactivity.

I'm looking how I can add a visual indicator

File: config/core.entity_form_display.node.training.default.yml
Comment (orakili) [inline config/core.entity_form_display.node.training.default.yml:200]:
```suggestion
    type: options_buttons
```

Not needed for the training format. It's only 2 choices:  online or on site and the description for those terms is basically the term label. It's also not required in the RW-601 ticket.

File: html/modules/custom/reliefweb_fields/components/reliefweb-options/reliefweb-options.js
Comment (orakili) [inline html/modules/custom/reliefweb_fields/components/reliefweb-options/reliefweb-options.js:143]: The problem with that approach is that it doesn't respect the maximum of allowed terms. For example by selecting the professional function terms in the popup, I could exceed the limit:

Screen Shot 2022-06-30 at 15 50 06

That's also going to cause issues with the `career categories` for jobs where the `theme` options become disabled for some career categories (for example themes are not allowed when selecting "Administration/Finance"):

Screen Shot 2022-06-30 at 15 53 21

I haven't tested but maybe another way to handle that could be to transform the original field wrapper into the popup. By that I mean, adding a class to the fieldset wrapper that would result in it being styled as a popup and showing the description via the a `::after` for example (a bit like the option 2 in the RW-601 ticket but with the field styled as a popup). If that works, that could reduce the complexity of the javascript as well.

File: html/modules/custom/reliefweb_fields/src/Plugin/Field/FieldWidget/ReliefWebOptions.php
Comment (orakili) [inline html/modules/custom/reliefweb_fields/src/Plugin/Field/FieldWidget/ReliefWebOptions.php:83]: I wonder if we can add a class to the field wrapper here to ease the selection in the javascript. Something like `$element['#wrapper_attributes']['class'] = 'data-with-term-descriptions';` (not tested).

Comment (attiks) [reply html/modules/custom/reliefweb_fields/src/Plugin/Field/FieldWidget/ReliefWebOptions.php:83]: `$element['#attributes']['class'][] = 'data-with-term-descriptions';` works

File: html/modules/custom/reliefweb_fields/components/reliefweb-options/reliefweb-options.js
Comment (attiks) [reply html/modules/custom/reliefweb_fields/components/reliefweb-options/reliefweb-options.js:143]: Fixed, attributes are copied and `Drupal.attachBehaviors(popup)` is called

**Review by orakili — CHANGES_REQUESTED** [review]
Sorry, there are still issues when selecting options in the popup.

For example, in the training form, if I select 3 themes in the popup and close it, the other options in the original field are not disabled which allows to select one more option (before the behavior kicks in and disables the remaining options).

On the **job** form, selecting "Administration/Finance" for example in the popup doesn't disable the theme field options as it should.

**Review by orakili — APPROVED** [review]
It seems to be working now, thanks. I'm approving but this needs to be styled before merging.

Also, for info, while testing I came across this issue: https://humanitarian.atlassian.net/browse/RW-605 which also happens on prod so it's related to this new feature.

**Comment by orakili** [comment]
@attiks For info, the fix in RW-605 (`document` -> `context`) broke the selection limit in the popup:

```
var elements = context.querySelectorAll('fieldset[data-with-selection-limit]:not([data-with-selection-limit-processed])');
```

because `context` is the `popup` element when calling `Drupal.attachBehaviors(popup)` and `popup` was the element with the `data-with-selection-limit` attribute.

So in https://github.com/UN-OCHA/rwint9-site/pull/389/commits/5267f5d38e434d0130ca570732ed08e15407d563, I added the attributes to the `container` element in the popup instead, to fix the issue.

I also changed back the `popup` to a `div` because it's not necessary to have it as a `fieldset` anymore and that will make it easier to style.

cc @left23 just a tiny ping to let you know that the popup most outer element is now a `div` instead of a `fieldset` like it is for the guidelines popups so it should be easier to re-use the styling from the `common_design_subtheme/components/rw-guidelines` component.

**Review by orakili — DISMISSED** [review]
Just re-approving as the last commit dismissed my previous review.

File: html/modules/custom/reliefweb_form/js/widget.selectionlimit.js
Comment (attiks) [inline html/modules/custom/reliefweb_form/js/widget.selectionlimit.js:74]: This means it will bind to all individual input elements as well, I just tried with the previous code and limit is working in page and popup :confused:

Comment (orakili) [reply html/modules/custom/reliefweb_form/js/widget.selectionlimit.js:74]: Ah, I didn't notice that the `data-with-selection-limit` was on the individual input as well as the checkboxes container. That was not the case in RW7 and that's also what the `checkSelectionLimit` function implies with `var inputs = element.getElementsByTagName('input');`.

Created https://humanitarian.atlassian.net/browse/RW-611 to solve the problem at the root and only have the attribute on the checkboxes container and not the individual inputs.

**Review by orakili — APPROVED** [review]

**Pull request merged**
